repository_service_tuf_worker/repository.py

```diff
@@ -287,87 +308,89 @@ def _update_snapshot(
         return snapshot.signed.version
 
     def _get_path_succinct_role(self, target_path: str) -> str:
+        """
+        Return role name by target file path
+        """
         bin_role = self._load(BIN)
         bin_succinct_roles = bin_role.signed.delegations.succinct_roles
         bins_name = bin_succinct_roles.get_role_for_target(target_path)
 
         return bins_name
 
-    def _add_to_unpublished_metas(self, targets_meta: List[Tuple[str, int]]):
-        """
-        Add to the Redis "unpublished_meta" the edited roles.
-        """
-        logging.debug(f"Adding to unpublished meta {targets_meta}")
-        if len(targets_meta) == 0:
-            logging.info("Nothing to send to be published")
-            return None
-
-        with self._redis.lock("TUF_TARGETS_META"):
-            if self._redis.exists("unpublished_metas"):
-                targets_waiting_commmit = self._redis.get(
-                    "unpublished_metas"
-                ).decode("utf-8")
-                for bins_name, _ in targets_meta:
-                    if bins_name not in targets_waiting_commmit:
-                        self._redis.append(
-                            "unpublished_metas", f", {bins_name}"
-                        )
-            else:
-                self._redis.set(
-                    "unpublished_metas",
-                    ", ".join(bins_name for bins_name, _ in targets_meta),
-                )
-
-    def _publish_meta_state(
-        self, targets_meta: List[Tuple[str, int]], update_state: Optional[str]
-    ) -> List[Optional[Tuple[str, int]]]:
+    def _update_task(
+        self,
+        bin_targets: Dict[str, List[targets_models.RSTUFTargets]],
+        update_state: Task.update_state,
+    ):
         """
         Updates the 'RUNNING' state with details if the meta still not
         published in the latest Snapshot. It runs every 3 seconds.
         """
-        logging.debug(f"waiting metas to be published {targets_meta}")
+        logging.debug(f"Waiting roles to be published {list(bin_targets)}")
 
-        def _update_state(targets_meta: List[Tuple[str, int]]):
-            unpublised_roles = [
-                f"{role} version {version}" for role, version in targets_meta
-            ]
+        def _update_state(
+            bin_targets: Dict[str, List[targets_models.RSTUFTargets]],
+            completed_roles: List[str],
+        ):
             update_state(
                 state="RUNNING",
                 meta={
-                    "unpublished_roles": unpublised_roles,
+                    "published_roles": completed_roles,
+                    "roles_to_publish": f"{list(bin_targets.keys())}",
                     "status": "Publishing",
                     "last_update": datetime.now(),
                 },
             )
 
         while True:
```

Comment:
Do you need the `while True` loop given that if the `completed_roles` list is different than the `bin_targets` list you will call the `_update_state` function anyway?

Reply:
Not clear if I understood the question.
Yes, I need it. Otherwise, the task initiated in the `add_targets` or `remove_targets` will finish.
The task needs to keep running until it finishes.

Reply:
What I mean here is that without the `while True` look you will go inside the
`for role_name, targets in bin_targets.items():`  and add some roles in `completed_roles`.
Then at line 357 you will check `if sorted(completed_roles) != sorted(list(bin_targets)):` and if that's true, then you will call update_state again and redo the same thing.
Not sure what is the benefit `while True` brings here as you will traverse the same code again if you are not finished.

Reply:
But I will run it only once, update_state once, return None and finish the task at line:

https://github.com/vmware/repository-service-tuf-worker/blob/38ea42cd64ea8b683b00e60da7f37b2f9d3dc806/repository_service_tuf_worker/repository.py#L542

or

https://github.com/vmware/repository-service-tuf-worker/blob/38ea42cd64ea8b683b00e60da7f37b2f9d3dc806/repository_service_tuf_worker/repository.py#L598

Before I finished all targets.

Reply:
I read the function again locally and it seems I read it incorrectly.
Thanks.